### PR TITLE
feat: Address scam badge flag

### PR DESCRIPTION
**Pull request by vbaranov**
Resolves https://github.com/blockscout/blockscout/issues/10677

## Motivation

Implementation of address badges functionality.

### Enhancements

- `is_scam` boolean flag to all API endpoints, which return address info.
- asset is removed from the search results, if its address has a scam badge, and `HIDE_SCAM_ADDRESSES=true` runtime variable is set.

Management API:

Screenshot 2024-09-13 at 16 07 11

[Blockscout API v2 address scam badges.postman_collection.json](https://github.com/user-attachments/files/16994421/Blockscout.API.v2.address.scam.badges.postman_collection.json)

Live examples:
https://eth-sepolia.k8s-dev.blockscout.com/api/v2/addresses/0x2a202d07f520f6aaee68d2dac9b2ef35834d5923
This address is currently removed from search results there.

## Checklist for your Pull Request (PR)

- [ ] If I added new functionality, I added tests covering it.
- [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
- [x] I checked whether I should update the docs and did so by submitting a PR to [docs repository](https://github.com/blockscout/docs).
- [x] If I added/changed/removed ENV var, I submitted a PR to [docs repository](https://github.com/blockscout/docs) to update the list of [env vars](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md) and I updated the version to `master` in the Version column. If I removed variable, I added it to [Deprecated ENV Variables](https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables/deprecated-env-variables/README.md) page. After merging docs PR, changes will be reflected in these [pages](https://docs.blockscout.com/for-developers/information-and-settings/env-variables). https://github.com/blockscout/docs/pull/326
- [x] If I added new DB indices, I checked, that they are not redundant, with PGHero or other tools.
- [x] If I added/removed chain type, I modified the Github CI matrix and PR labels accordingly.

### CODE REVIEW

File: apps/block_scout_web/lib/block_scout_web/routers/api_router.ex [hunk 1]
```diff
@@ -57,6 +66,23 @@ defmodule BlockScoutWeb.Routers.ApiRouter do
     plug(RateLimit)
   end
 
+  pipeline :api_v2_no_forgery_protect do
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/routers/api_router.ex:69]: is this pipeline used somewhere?

File: apps/block_scout_web/.sobelow-conf [hunk 2]
```diff
@@ -9,6 +9,8 @@
   ignore_files: [
     "apps/block_scout_web/lib/block_scout_web/routers/smart_contracts_api_v2_router.ex",
     "apps/block_scout_web/lib/block_scout_web/routers/tokens_api_v2_router.ex",
-    "apps/block_scout_web/lib/block_scout_web/routers/utils_api_v2_router.ex"
+    "apps/block_scout_web/lib/block_scout_web/routers/utils_api_v2_router.ex",
+    "apps/block_scout_web/lib/block_scout_web/routers/address_badges_v2_router.ex",
+    "apps/block_scout_web/lib/block_scout_web/routers/api_router.ex"
```
Comment (nikitosing) [inline apps/block_scout_web/.sobelow-conf:14]: is it debug artifact? (main `api_router` in sobelow-ignore)

File: apps/block_scout_web/lib/block_scout_web/controllers/account/api/v2/fallback_controller.ex [hunk 3]
```diff
@@ -85,7 +85,6 @@ defmodule BlockScoutWeb.Account.Api.V2.FallbackController do
 
   def call(conn, {:api_key, _}) do
     conn
-    |> put_status(:unauthorized)
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/controllers/account/api/v2/fallback_controller.ex:88]: why do you think that http code 200 better than 403 here?

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+         {:api_key, ^api_key} <- {:api_key, params["api_key"]},
+         {_num_of_inserted, badge_to_address_list} <- ScamBadgeToAddress.add(address_hashes) do
+      conn
+      |> put_status(200)
+      |> render(:badge_to_address, %{
+        badge_to_address_list: badge_to_address_list,
+        status: if(Enum.empty?(badge_to_address_list), do: "update skipped", else: "added")
+      })
+    else
+      {:error, error} ->
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex:30]: Seems to me, that this clause is unreachable

Comment (nikitosing) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex:30]: in other functions too

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  end
+
+  def unassign_badge_from_address(_, _), do: {:error, :not_found}
+
+  def show_badge_addresses(conn, _) do
+    with {:ok, body, _conn} <- Conn.read_body(conn, []),
+         {:ok, %{"api_key" => provided_api_key}} <- Jason.decode(body),
+         {:sensitive_endpoints_api_key, api_key} when not is_nil(api_key) <-
+           {:sensitive_endpoints_api_key, Application.get_env(:block_scout_web, :sensitive_endpoints_api_key)},
+         {:api_key, ^api_key} <- {:api_key, provided_api_key} do
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex:75]: lets extract all such checks to a separate module in order to reuse and avoid code duplications.
I'd suggest here replace this:
```
{:sensitive_endpoints_api_key, api_key} when not is_nil(api_key) <-
           {:sensitive_endpoints_api_key, Application.get_env(:block_scout_web, :sensitive_endpoints_api_key)},
{:api_key, ^api_key} <- {:api_key, provided_api_key}
```
to this:
```
:ok <- Helper.check_sensitive_enpoint_api_key(api_key)
```

add BlockScoutWeb.Helper
and in BlockScoutWeb.Helper add:
```
def check_sensitive_enpoint_api_key(provided_api_key) do
  with {:sensitive_endpoints_api_key, api_key} when not is_nil(api_key) <-
           {:sensitive_endpoints_api_key, Application.get_env(:block_scout_web, :sensitive_endpoints_api_key)},
      {:api_key, ^api_key} <- {:api_key, provided_api_key} do
    :ok
  end
```

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      _ ->
+        {:api_key, :wrong}
+    end
+  end
+
+  def unassign_badge_from_address(_, _), do: {:error, :not_found}
+
+  def show_badge_addresses(conn, _) do
+    with {:ok, body, _conn} <- Conn.read_body(conn, []),
+         {:ok, %{"api_key" => provided_api_key}} <- Jason.decode(body),
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex:72]: mb it's better to swtich method to POST or expect api_key in header?

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        status: if(Enum.empty?(badge_to_address_list), do: "update skipped", else: "removed")
+      })
+    else
+      {:error, error} ->
+        Logger.error(fn -> ["Badge addresses addition failed: ", inspect(error)] end)
+        {:error, :badge_creation_failed}
+
+      _ ->
+        {:api_key, :wrong}
+    end
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex:65]:
```suggestion
    end
```

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        status: if(Enum.empty?(badge_to_address_list), do: "update skipped", else: "added")
+      })
+    else
+      {:error, error} ->
+        Logger.error(fn -> ["Badge addresses addition failed: ", inspect(error)] end)
+        {:error, :badge_creation_failed}
+
+      _ ->
+        {:api_key, :wrong}
+    end
```
Comment (nikitosing) [inline apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex:36]:
```suggestion
    end
```

File: apps/block_scout_web/lib/block_scout_web/routers/api_router.ex (hunk 1 above)
Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/routers/api_router.ex]: This is a temporary development artefact. Thanks. I removed it in https://github.com/blockscout/blockscout/pull/10763/commits/bb25a53446fc5bc6ddb9e63802fe97272947e138.

File: apps/block_scout_web/.sobelow-conf (hunk 2 above)
Comment (vbaranov) [reply apps/block_scout_web/.sobelow-conf]: yep, removed in This is a temporary development artefact. I removed it in https://github.com/blockscout/blockscout/pull/10763/commits/bb25a53446fc5bc6ddb9e63802fe97272947e138..

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex (hunk 5 above)
Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/address_badge_controller.ex]: Done in https://github.com/blockscout/blockscout/pull/10763/commits/bb25a53446fc5bc6ddb9e63802fe97272947e138.

File: apps/block_scout_web/lib/block_scout_web/controllers/account/api/v2/fallback_controller.ex (hunk 3 above)
Comment (vbaranov) [reply apps/block_scout_web/lib/block_scout_web/controllers/account/api/v2/fallback_controller.ex]: That was an unintentional change. I removed it.

**Review by nikitosing — APPROVED** [review]

**Pull request merged**